core/src/main/java/io/grpc/internal/DelayedClientTransport.java

```diff
@@ -270,48 +308,61 @@ final int getPendingStreamsCount() {
+    ArrayList<PendingStream> newlyCreated = new ArrayList<>();
 
-    for (final PendingStream stream : toProcess) {
-      PickResult pickResult = picker.pickSubchannel(stream.args);
-      CallOptions callOptions = stream.args.getCallOptions();
-      final ClientTransport transport = GrpcUtil.getTransportFromPickResult(pickResult,
-          callOptions.isWaitForReady());
-      if (transport != null) {
-        Executor executor = defaultAppExecutor;
-        // createRealStream may be expensive. It will start real streams on the transport. If
-        // there are pending requests, they will be serialized too, which may be expensive. Since
-        // we are now on transport thread, we need to offload the work to an executor.
-        if (callOptions.getExecutor() != null) {
-          executor = callOptions.getExecutor();
-        }
-        executor.execute(new Runnable() {
+
+    if (picker != null) {
+      for (final PendingStream stream : toCreateRealStream) {
+        PickResult pickResult = picker.pickSubchannel(stream.args);
+        CallOptions callOptions = stream.args.getCallOptions();
+        final ClientTransport transport = GrpcUtil.getTransportFromPickResult(pickResult,
+            callOptions.isWaitForReady());
+        if (transport != null) {
+          Executor executor = defaultAppExecutor;
+          // createRealStream may be expensive. It will start real streams on the transport. If
+          // there are pending requests, they will be serialized too, which may be expensive. Since
+          // we are now on transport thread, we need to offload the work to an executor.
+          if (callOptions.getExecutor() != null) {
+            executor = callOptions.getExecutor();
+          }
+          executor.execute(new Runnable() {
             @Override
             public void run() {
               stream.createRealStream(transport);
             }
           });
-        toRemove.add(stream);
-      }  // else: stay pending
+          newlyCreated.add(stream);
+        }  // else: stay pending
+      }
+    }
+    toCheckCompletion.addAll(newlyCreated);
+    ArrayList<PendingStream> completed = new ArrayList<>();
+    for (final PendingStream stream : toCheckCompletion) {
+      if (stream.isStreamTransferCompleted()) {
+        completed.add(stream);
+      }
     }
-
     synchronized (lock) {
       // Between this synchronized and the previous one:
       //   - Streams may have been cancelled, which may turn pendingStreams into emptiness.
-      //   - shutdown() may be called, which may turn pendingStreams into null.
-      if (!hasPendingStreams()) {
+      //   - shutdownNow() may be called, which may turn pendingStreams into emptiness.
+      if (!hasPendingStreams() && !hasUncommittedStreams()) {
         return;
       }
-      pendingStreams.removeAll(toRemove);
+      pendingStreams.removeAll(newlyCreated);
+      toCheckCompletionStreams.addAll(newlyCreated);
```

Comment:
I would prefer making `toCheckCompletionStreams` be either 'non-final and immutable' or 'final and mutable', but not 'non-final' and 'mutable'.